Task 1723: VM experiments (#81)
SecConNet/mahiru: docs/distributed_scenarios.rst

> Running distributed scenarios

Reviewer: I love this documentation Lourens! It's really clear, and necessary as well. I think it is a good idea to test whether someone (me?) can reproduce this. I've openend #84 for that.

Author: Yeah. A future me will probably be doing that, and by then I'll have forgotten how this works, so that will be a reasonable simulation, but an actual other person would be even better. Thanks!

> Running distributed scenarios

Reviewer: I take it from you adding these instructions that this scenario is intended to be used by others right? At some places it feels a bit like: 'I've done this experiment, I hope it also works for you, good luck!'. I think it's really subtle, but you could try rewriting it a bit so it feels more like we expect loads and loads of users :) i.e. that it reads more like: 'This is how you should run this experiment'

Reviewer: I added a few suggestions to change the tone a bit, curious what you think of it!

Author: "I've done this experiment, I hope it also works for you, good luck!" is a pretty good description of the state this software is in, in my perception. I'll come back to this documentation once I actually have some decent unit test coverage and error handling and rephrase things here and there, but for now that would violate truth in advertising regulations ;-).

Reviewer: OK, I think I understand your motivation for writing it this way. I still disagree ;) I'm exaggerating, but the way you see it is that a user-friendly (in the sense of inviting people to use the software instead of scaring away) immediately promises a finished, reliable product. You can also have user-friendly documentation and have a disclaimer somewhere that it is still a prototype. The code quality and documentation of mahiru is much better than some software out there that people use everyday in production.

Reviewer: This is my reply also to the other comments, and I think this is a huge improvement over the old code, so feel free to merge! 🚀

=============================

Mahiru is a distributed system. While that doesn't stop anyone from running a
complete Mahiru system on a single machine, some experiments require setting up
Mahiru on actual different machines to get realistic results.

So, we need some way of running distributed experiments: we obtain some
(physical or virtual) machines, install a Mahiru registry on one of them and
some sites on others, and run some kind of distributed processing scenario.
Mahiru has a set-up for this, and this page explains how to work with it.

Installing software
-------------------

The distributed set-up uses `Vagrant <https://vagrantup.com>`_ and `Ansible
<https://ansible.com>`_ to orchestrate the experiment, and we'll also need
`Docker <https://docker.com>`_ to build some container images. I'm running this
on Ubuntu Linux, so that's what these instructions are for. If you have a
different OS, then you'll probably have to do some figuring-out to get things to

> work.

Reviewer: Can we make this a bit more non-linux-user friendly? Maybe add some pointers on where to start (i.e. point to the troubleshooting at the end of the file). And add something like 'If you are stuck, please file an issue so we can help you, see CONTRIBUTING.md'

Author: My intended audience for now is computer science PhD students who have a significant skill set, who are willing to invest some time in figuring things out, and whom I'm probably working with directly. The software quality of this thing isn't anywhere near the point where I would want to start offering support, and then there's the issue that I know nothing at all about virtual machines on MacOS or Windows, so I cannot really help anyway.

So I agree in principle, and the MUSCLE3 documentation has lots of that kind of stuff, but here and now "you'll probably have to do some figuring-out" is just a realistic description of the situation the reader is in...

Vagrant creates virtual machines according to a desired configuration and takes
care of setting up access via SSH. Ansible configures virtual or physical
machines by connecting to them, examining their state, and taking action to
ensure it matches some desired state described in its configuration. Vagrant can
call Ansible to configure (or *provision*) the VMs it has created. Docker is
used both to deploy Mahiru on the machines and by Mahiru itself to execute
workflow steps.

> If you use Vagrant, then the experiment will be set up locally on a set of
> virtual machines. If you have virtual or physical machines available already,
> then you can skip Vagrant and instead tell Ansible about them, then use Ansible
> directly to install the software and run the experiment.

Reviewer: ```suggestion
We recommend using Vagrant, then the experiment will be set up locally on a set of
virtual machines. For advanced users, if you have virtual or physical machines available already,
then you can skip Vagrant and instead tell Ansible about them, then use Ansible
directly to install the software and run the experiment.
```

Author: Same here. I'm actually not explaining how to do this with bare Ansible, without Vagrant. I may try this on the UvA physical machines at some point, so I'll figure that out then and add it.

> I'm running this locally, with Vagrant and Ansible, and using ``libvirt`` with

Reviewer: ```suggestion
We recommend running this locally, with Vagrant and Ansible, and using ``libvirt`` with
```

Author: I don't really want to recommend Vagrant :-). It also depends on what you want to do, if you want to measure network throughput in the system, then I don't recommend doing that on one machine with Vagrant-launched VMs, because that would not be very realistic. So it depends quite a bit on what the user wants to do.

``VAGRANT_DEFAULT_PROVIDER`` to ``libvirt`` for it to automatically talk to the
correct hypervisor.

Installing and configuring Ansible
``````````````````````````````````

Ansible is written in Python, and can be installed using ``pip``:

.. code-block:: bash

  pip3 install ansible

More information on installing Ansible can be found in the `Ansible Installation
Guide <https://docs.ansible.com/ansible/latest/installation_guide/index.html>`_.

Configuring machines with Ansible can be rather slow, but there are some
settings that can be changed to improve that, and it's well worth doing so. To
do this, create a file in your home directory called `.ansible.cfg` and add the
following content:

.. code-block:: ini

  [defaults]
  use_persistent_connections = True
  ssh_args = -o ControlMaster=auto -o ControlPersist=36000s
  control_path = %(directory)s/ansible-ssh-%%h-%%p-%%r
  pipelining = True

  deprecation_warnings = False

These settings make Ansible connect to the machine that is being configured
once, then reuse the connection to do multiple things. This is quite a bit
faster than reconnecting all the time, because SSH has a lot of connection
set-up overhead.

> With my particular combination of Vagrant and Ansible, I was getting some
> deprecation warnings. Everything seems to work fine however, so I disabled them
> here to cut down on the clutter.

Reviewer: ```suggestion
With some combinations of Vagrant and Ansible, you can get
deprecation warnings. Everything seems to work fine however, so 
we disable them here to cut down on the clutter.
```

install it from there. However, it does pay to have a recent version of Docker
with support for Buildkit, because that lets you build images much more quickly.

Building container images
-------------------------

We'll be running the Mahiru registry and Mahiru sites on our machines, and we're
going to use Docker to deploy it. To make that work, we need to build some
Docker images containing Mahiru. This is done using ``make`` in the root of the
repository:

.. code-block:: bash

  make docker_tars

(If you don't have ``make`` installed, this will give an error. ``sudo apt-get
install make`` should fix that.)

Mahiru runs workflows, and each step in a workflow executes a program or script.
To allow sending these programs to other sites and running them there, they are
packaged up into container images as well. The same goes for data sets. Mahiru
calls these programs and data sets *assets*. To run our experiment, we'll need
some of these asset container images. They can be built in the same way:

.. code-block:: bash

  make assets

The container images this produces can be found in the `build/` directory, if

> you're interested (but honestly, there's not much to see).

Reviewer: Lol

``client2``. This runs a site-specific script on each client machine which
connects to the corresponding site and sets it up to be able to run a simple
compute-to-data scenario.

Finally, we can submit a processing workflow to ``site1`` via ``client1``. Due
to the way the policies are set up, it will be executed in a compute-to-data
way:

.. code-block:: bash

  vagrant provision --provision-with run_experiment client1

The typical workflow after making changes to Mahiru or to the experiment is to
first shut down Mahiru, then restart it with the new code, and then repeat
``configure_sites`` and ``run_experiment`` as above. Stopping and starting
Mahiru without destroying and recreating the VMs can be done with

.. code-block:: bash

  vagrant provision --provision-with mahiru_down registry client1 client2
  vagrant provision --provision-with mahiru_up registry client1 client2

To destroy the VMs, use (without -f you'll have to confirm the destruction of
each individual VM):

.. code-block:: bash

  vagrant destroy -f

How it works
------------

This section is a bit short for now, but hopefully provides a bit of a start to

> figuring out how all this is put together.

Reviewer: I think it's already at 90%, but good to give this disclaimer.
